smelter-core/src/pipeline/v4l2/v4l2_input.rs

```diff
@@ -0,0 +1,260 @@
+            InputInitInfo::Other,
+            QueueDataReceiver {
+                video: Some(rx),
+                audio: None,
+            },
+        ))
+    }
+}
+
+impl Drop for V4l2Input {
+    fn drop(&mut self) {
+        self.should_close
+            .store(true, std::sync::atomic::Ordering::Relaxed);
+    }
+}
+
+struct InputState<'a> {
+    v4l_device: Device,
+    resolution: Resolution,
+    format: V4l2Format,
+    ctx: Arc<PipelineCtx>,
+    should_close: Arc<AtomicBool>,
+    sender: crossbeam_channel::Sender<PipelineEvent<Frame>>,
+    stream: v4l::io::mmap::Stream<'a>,
+}
+
+impl InputState<'_> {
+    fn run(&mut self) {
+        loop {
+            if self.should_close.load(std::sync::atomic::Ordering::Relaxed) {
+                self.send_eos();
+                return;
+            }
+
+            let frame = match self.stream.next() {
+                Ok((frame, _)) => frame,
+                Err(e) => {
+                    warn!(error = %e, "Cannot receive frame.");
+                    continue;
+                }
+            };
+
+            let data = match self.format {
+                V4l2Format::Yuyv => {
+                    if frame.len() != self.resolution.width * self.resolution.height * 2 {
+                        if let Err(e) = self.refresh_format() {
+                            error!(error= %e, "Fatal error when trying to refresh parameters.");
+                            self.send_eos();
+                            return;
+                        }
+
+                        continue;
+                    }
+
+                    smelter_render::FrameData::InterleavedYuyv422(bytes::Bytes::copy_from_slice(
+                        frame,
+                    ))
+                }
+                V4l2Format::Nv12 => {
+                    let y_length = self.resolution.width * self.resolution.height;
+                    if frame.len() != y_length * 3 / 2 {
+                        if let Err(e) = self.refresh_format() {
+                            error!(error = %e, "Fatal error when trying to refresh parameters.");
+                            self.send_eos();
+                            return;
+                        }
+
+                        continue;
+                    }
+
+                    smelter_render::FrameData::Nv12(smelter_render::NvPlanes {
+                        y_plane: bytes::Bytes::copy_from_slice(&frame[..y_length]),
+                        uv_planes: bytes::Bytes::copy_from_slice(&frame[y_length..]),
+                    })
+                }
+            };
+
+            let frame = Frame {
+                pts: self.ctx.queue_sync_point.elapsed() + std::time::Duration::from_millis(20),
+                data,
+                resolution: self.resolution,
+            };
+
+            if let Err(e) = self.sender.send(PipelineEvent::Data(frame)) {
+                debug!(error = %e, "Failed to send video chunk.");
```

Comment:
```suggestion
                debug!("Failed to send video frame from V4L2 input. Channel closed.");
```
In other places we don't print the error if we know that it is channel closed